[CSS Zoom] Apply zoom factor to margin

brentfulgham opened the pull request:
#### f028d47877de5efb726d36856638e43c429625cf
<pre>
[CSS Zoom] Apply zoom factor to margin
<a href="https://bugs.webkit.org/show_bug.cgi?id=301024">https://bugs.webkit.org/show_bug.cgi?id=301024</a>
<a href="https://rdar.apple.com/162907254">rdar://162907254</a>

Reviewed by Vitor Roriz.

Adapt margin logic to the new CSS Zoom architecture.

Tests: imported/w3c/web-platform-tests/css/css-viewport/zoom/list-style-image.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/margin.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html

* LayoutTests/TestExpectations
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/margin-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/margin.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/margin-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::pageSizeAndMarginsInPixels):
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::computedHorizontalMargin const):
(WebCore::Layout::FormattingGeometry::computedVerticalMargin const):
* Source/WebCore/layout/formattingContexts/FormattingGeometry.h:
(WebCore::Layout::FormattingGeometry::computedValue const):
(WebCore::Layout::FormattingGeometry::fixedValue const):
* Source/WebCore/layout/formattingContexts/block/BlockFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/block/BlockFormattingGeometry.cpp:
(WebCore::Layout::BlockFormattingGeometry::intrinsicWidthConstraints const):
* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::layout):
(WebCore::Layout::GridLayout::computeInlineMargins):
(WebCore::Layout::GridLayout::computeBlockMargins):
* Source/WebCore/layout/formattingContexts/grid/GridLayout.h:
* Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp:
(WebCore::LayoutIntegration::usedValueOrZero):
(WebCore::LayoutIntegration::BoxGeometryUpdater::horizontalLogicalMargin):
(WebCore::LayoutIntegration::BoxGeometryUpdater::verticalLogicalMargin):
* Source/WebCore/layout/Verification.cpp:
(WebCore::Layout::outputMismatchingBlockBoxInformationIfNeeded):
* Source/WebCore/page/PrintContext.cpp:
(WebCore::PrintContext::computedPageMargin):
(WebCore::PrintContext::pageProperty):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::setup):
* Source/WebCore/rendering/PositionedLayoutConstraints.h:
(WebCore::PositionedLayoutConstraints::marginBeforeValue const):
(WebCore::PositionedLayoutConstraints::marginAfterValue const):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::marginIntrinsicLogicalWidthForChild const):
(WebCore::RenderBlock::computeBlockPreferredLogicalWidths const):
(WebCore::RenderBlock::computePreferredWidthsForExcludedChildren const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::getBorderPaddingMargin):
(WebCore::RenderBlockFlow::computeInlinePreferredLogicalWidths const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computeLogicalWidth const):
(WebCore::RenderBox::fillAvailableMeasure const):
(WebCore::RenderBox::computeInlineDirectionMargins const):
(WebCore::RenderBox::constrainBlockMarginInAvailableSpaceOrTrim const):
* Source/WebCore/rendering/RenderBoxInlines.h:
(WebCore::RenderBox::marginBoxRect const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::computeStickyPositionConstraints const):
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
(WebCore::RenderBoxModelObject::resolveLengthPercentageUsingContainerLogicalWidth const):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::marginWidthForChild):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeFlexItemMarginValue):
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::computeMargin):
(WebCore::RenderInline::marginLeft const):
(WebCore::RenderInline::marginRight const):
(WebCore::RenderInline::marginTop const):
(WebCore::RenderInline::marginBottom const):
(WebCore::RenderInline::marginStart const):
(WebCore::RenderInline::marginEnd const):
(WebCore::RenderInline::marginBefore const):
(WebCore::RenderInline::marginAfter const):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::layout):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/RenderScrollbarPart.cpp:
(WebCore::RenderScrollbarPart::computeScrollbarWidth):
(WebCore::RenderScrollbarPart::computeScrollbarHeight):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::updateLogicalWidth):
* Source/WebCore/rendering/RenderTextInlines.h:
(WebCore::RenderText::marginLeft const):
(WebCore::RenderText::marginRight const):
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::ExtractorCustom::extractMarginRight):
(WebCore::Style::ExtractorCustom::extractMarginRightSerialization):
* Source/WebCore/style/values/box/StyleMargin.cpp: Added.
* Source/WebCore/style/values/box/StyleMargin.h:

Canonical link: <a href="https://commits.webkit.org/301965@main">https://commits.webkit.org/301965@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/16382139fd81318021b975daee7f4bf091420e99

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127615 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47263 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134886 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/79172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/80dd138d-19b2-442d-85d6-6bffb8acf349) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129487 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47886 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55793 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97145 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/79172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5c16a235-2a59-49df-81c0-1414bde1a461) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38298 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114313 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77626 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c0d4c183-7496-4c1d-aeb6-f4e475f7ed3d) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/37106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32403 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78245 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108166 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137367 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54274 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41831 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105673 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54785 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110670 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105324 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26857 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50836 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51868 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54211 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60379 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53445 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56902 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55204 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nullhook commented on Source/WebCore/style/values/box/StyleMargin.cpp:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    state.setCurrentPropertyInvalidAtComputedValueTime();
+    return CSS::Keyword::Auto { };
```
after you create the custom conversionData, the rest of the logic in `MarginEdge` looks identical to the generic LengthWrapperBase conversion.

probably can extract it to a generic helper

```
template<LengthWrapperBaseDerived T>
auto convertLengthPercentageWithConversionData(BuilderState& state, const CSSPrimitiveValue& primitiveValue, const CSSToLengthConversionData& conversionData) -> T {}
```

and for return types we can perhaps do some `T::Keywords` check here or fallback to `0_css_px;` for types without keywords.

`MarginEdge` can use `convertLengthPercentageWithConversionData<MarginEdge>()`, and `WordSpacing` can also use it, along with the generic path.

it’s probably not too important for this PR, but something to think about for the future.

nullhook commented on Source/WebCore/style/values/box/StyleMargin.cpp:
```diff
@@ -0,0 +1,93 @@
+        return CSS::Keyword::Auto { };
+
+    if (primitiveValue->valueID() == CSSValueAuto)
+        return CSS::Keyword::Auto { };
```
you can collapse this to `if (!primitiveValue || primitiveValue->valueID() == CSSValueAuto) {}`

brentfulgham replied:
Will do!

nullhook commented on Source/WebCore/style/values/box/StyleMargin.cpp:
```diff
@@ -0,0 +1,93 @@
+            usedZoom = zoomWithTextZoomFactor(state);
+
+        if (usedZoom == zoom)
+            return state.cssToLengthConversionData();
```
can you comment on why do we need this for MarginEdge?

brentfulgham replied:
I copied this from Vitor's change, assuming it was a perf benefit (if our calculated zoom already matches the state's Length-adjusted zoom, just return it rather than adjusting to the same zoom.

vitorroriz commented on Source/WebCore/style/values/box/StyleMargin.cpp:
```diff
@@ -0,0 +1,95 @@
+    auto conversionData = state.useSVGZoomRulesForLength()
+        ? state.cssToLengthConversionData().copyWithAdjustedZoom(1.0f)
+        : cssToLengthConversionDataWithTextZoomFactorIfNeeded(state, primitiveValue.isFontIndependentLength());
+
+    if (primitiveValue.isLength()) {
+        return MarginEdge {
+            typename MarginEdge::Fixed {
+                CSS::clampToRange<MarginEdge::Fixed::range, float>(primitiveValue.resolveAsLength(conversionData), minValueForCssLength, maxValueForCssLength),
+            },
+            primitiveValue.primitiveType() == CSSUnitType::CSS_QUIRKY_EM
+        };
+    }
```
From my investigation, the issue here is similar to when resolving LineHeight to a fixed value that with a font-relative unit. Although normally zoom would not affect the computed size of the property the current behavior is that the computed size in that case scales with text zoom, specifically. So we need something like this. 
```suggestion
    if (primitiveValue.isLength()) {
        auto textZoom = (evaluationTimeZoomEnabled(state) && !primitiveValue.isFontIndependentLength()) ? conversionData.zoom() : 1.0f;
        return MarginEdge {
            typename MarginEdge::Fixed {
                CSS::clampToRange<MarginEdge::Fixed::range, float>(primitiveValue.resolveAsLength(conversionData) * textZoom, minValueForCssLength, maxValueForCssLength),
            },
            primitiveValue.primitiveType() == CSSUnitType::CSS_QUIRKY_EM
        };
    }
```

brentfulgham committed:
[CSS Zoom] Apply zoom factor to margin
https://bugs.webkit.org/show_bug.cgi?id=301024
rdar://162907254

Reviewed by Vitor Roriz.

Adapt margin logic to the new CSS Zoom architecture.

Tests: imported/w3c/web-platform-tests/css/css-viewport/zoom/list-style-image.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/margin.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html

* LayoutTests/TestExpectations
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/margin-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/margin.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/margin-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::pageSizeAndMarginsInPixels):
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::computedHorizontalMargin const):
(WebCore::Layout::FormattingGeometry::computedVerticalMargin const):
* Source/WebCore/layout/formattingContexts/FormattingGeometry.h:
(WebCore::Layout::FormattingGeometry::computedValue const):
(WebCore::Layout::FormattingGeometry::fixedValue const):
* Source/WebCore/layout/formattingContexts/block/BlockFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/block/BlockFormattingGeometry.cpp:
(WebCore::Layout::BlockFormattingGeometry::intrinsicWidthConstraints const):
* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::layout):
(WebCore::Layout::GridLayout::computeInlineMargins):
(WebCore::Layout::GridLayout::computeBlockMargins):
* Source/WebCore/layout/formattingContexts/grid/GridLayout.h:
* Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp:
(WebCore::LayoutIntegration::usedValueOrZero):
(WebCore::LayoutIntegration::BoxGeometryUpdater::horizontalLogicalMargin):
(WebCore::LayoutIntegration::BoxGeometryUpdater::verticalLogicalMargin):
* Source/WebCore/layout/Verification.cpp:
(WebCore::Layout::outputMismatchingBlockBoxInformationIfNeeded):
* Source/WebCore/page/PrintContext.cpp:
(WebCore::PrintContext::computedPageMargin):
(WebCore::PrintContext::pageProperty):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::setup):
* Source/WebCore/rendering/PositionedLayoutConstraints.h:
(WebCore::PositionedLayoutConstraints::marginBeforeValue const):
(WebCore::PositionedLayoutConstraints::marginAfterValue const):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::marginIntrinsicLogicalWidthForChild const):
(WebCore::RenderBlock::computeBlockPreferredLogicalWidths const):
(WebCore::RenderBlock::computePreferredWidthsForExcludedChildren const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::getBorderPaddingMargin):
(WebCore::RenderBlockFlow::computeInlinePreferredLogicalWidths const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computeLogicalWidth const):
(WebCore::RenderBox::fillAvailableMeasure const):
(WebCore::RenderBox::computeInlineDirectionMargins const):
(WebCore::RenderBox::constrainBlockMarginInAvailableSpaceOrTrim const):
* Source/WebCore/rendering/RenderBoxInlines.h:
(WebCore::RenderBox::marginBoxRect const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::computeStickyPositionConstraints const):
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
(WebCore::RenderBoxModelObject::resolveLengthPercentageUsingContainerLogicalWidth const):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::marginWidthForChild):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeFlexItemMarginValue):
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::computeMargin):
(WebCore::RenderInline::marginLeft const):
(WebCore::RenderInline::marginRight const):
(WebCore::RenderInline::marginTop const):
(WebCore::RenderInline::marginBottom const):
(WebCore::RenderInline::marginStart const):
(WebCore::RenderInline::marginEnd const):
(WebCore::RenderInline::marginBefore const):
(WebCore::RenderInline::marginAfter const):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::layout):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/RenderScrollbarPart.cpp:
(WebCore::RenderScrollbarPart::computeScrollbarWidth):
(WebCore::RenderScrollbarPart::computeScrollbarHeight):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::updateLogicalWidth):
* Source/WebCore/rendering/RenderTextInlines.h:
(WebCore::RenderText::marginLeft const):
(WebCore::RenderText::marginRight const):
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::ExtractorCustom::extractMarginRight):
(WebCore::Style::ExtractorCustom::extractMarginRightSerialization):
* Source/WebCore/style/values/box/StyleMargin.cpp: Added.
* Source/WebCore/style/values/box/StyleMargin.h:

Canonical link: https://commits.webkit.org/301965@main
